src/pybind/mgr/telegraf/module.py

```diff
@@ -0,0 +1,315 @@
+import errno
+import json
+import socket
+import sys
+import time
+from threading import Event
+
+from telegraf.basesocket import BaseSocket
+from telegraf.protocol import Line
+from mgr_module import MgrModule
+
+if sys.version_info >= (3, 0):
+    from urllib.parse import urlparse
+else:
+    from urlparse import urlparse
+
+
+class Module(MgrModule):
+    COMMANDS = [
+        {
+            "cmd": "telegraf config-set name=key,type=CephString "
+                   "name=value,type=CephString",
+            "desc": "Set a configuration value",
+            "perm": "rw"
+        },
+        {
+            "cmd": "telegraf config-show",
+            "desc": "Show current configuration",
+            "perm": "r"
+        },
+        {
+            "cmd": "telegraf send",
+            "desc": "Force sending data to Telegraf",
+            "perm": "rw"
+        },
+        {
+            "cmd": "telegraf self-test",
+            "desc": "debug the module",
+            "perm": "rw"
+        },
+    ]
+
+    OPTIONS = [
+        {
+            'name': 'address',
+            'default': 'unixgram:///tmp/telegraf.sock',
+        },
+        {
+            'name': 'interval',
+            'default': 15
+        }
+    ]
+
+    fsid = None
+    run = False
+    ceph_health_mapping = {'HEALTH_OK': 0, 'HEALTH_WARN': 1, 'HEALTH_ERR': 2}
+
+    @property
+    def config_keys(self):
+        return dict((o['name'], o.get('default', None)) for o in self.OPTIONS)
+
+    def __init__(self, *args, **kwargs):
+        super(Module, self).__init__(*args, **kwargs)
+        self.event = Event()
+        self.run = True
+        self.config = dict()
+
+    def get_fsid(self):
+        if not self.fsid:
+            self.fsid = self.get('mon_map')['fsid']
+
+        return self.fsid
+
+    def get_pool_stats(self):
+        df = self.get('df')
+        data = []
+
+        df_types = [
+            'bytes_used',
+            'kb_used',
+            'dirty',
+            'rd',
+            'rd_bytes',
+            'raw_bytes_used',
+            'wr',
+            'wr_bytes',
+            'objects',
+            'max_avail',
+            'quota_objects',
+            'quota_bytes'
+        ]
+
+        for df_type in df_types:
+            for pool in df['pools']:
+                point = {
+                    'measurement': 'ceph_pool_stats',
+                    'tags': {
+                        'pool_name': pool['name'],
+                        'pool_id': pool['id'],
+                        'type_instance': df_type,
+                        'fsid': self.get_fsid()
+                    },
+                    'value': pool['stats'][df_type],
+                }
+                data.append(point)
+        return data
+
+    def get_daemon_stats(self):
```

Comment:
Can we share all these common bits between this module and the influx module?  Even though they're loaded completely separately, it should still work to do a "from influx.module import Foo", if the common bits were separated into a class Foo.

Reply:
I understand your reasoning. I've given this some thought, but as there was no stable way to do this.

The code seems similar between Influx and Telegraf, but I've made slight changes to those functions in the Telegraf module.

At some point we might want to have a "common" or "tools" in the Mgr I think.